### PR TITLE
New logging support to "Trainer" Class (ClearML Logger)

### PR DESCRIPTION
**Pull request by skinan**
I have added a ClearML callback class to log experiments using `ClearML Task.` ClearML logger logs everything to ClearML WebUI. ClearML logs Hyperparameters, Scalars, Models, Checkpoints, and other necessary artifacts.  I request @sgugger  to review this pull request as it works with the `Trainer` class. The `integrations.py` contains the major contents of this pull request through the `ClearMLCallback` class. Some other small changes have been made to another set of files to maintain the integrity of the codebase, such as adding simple basic tests.

> ClearML is a leading MLOps stack that can supercharge HuggingFace Transformers training and tracking with its state-of-the-art experiment tracking capability. ClearML: https://clear.ml/

**What ClearML Experiment Manager can log? Everything! You just name it. Example Screenshots:**
Screenshot from 2022-11-12 21-28-41

Screenshot from 2022-11-12 21-29-10

Screenshot from 2022-11-12 21-27-27
Screenshot from 2022-11-12 21-28-13
Screenshot from 2022-11-12 22-07-02

Screenshot from 2022-11-12 21-26-58
Screenshot from 2022-11-12 22-06-34

Screenshot from 2022-11-12 21-28-27
### 
**Example script to utilize ClearML Callback with Trainer:**

[trainer_with_clearml.zip](https://github.com/huggingface/transformers/files/9995390/trainer_with_clearml.zip)

### CODE REVIEW

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -1299,6 +1305,114 @@ def on_train_end(self, args, state, control, **kwargs):
             self.tracker.stop()
 
 
+class ClearMLCallback(TrainerCallback):
+    """
+    A [`TrainerCallback`] that sends the logs to [ClearML](https://clear.ml/).
+
+    Environment:
+            CLEARML_PROJECT (`str`, *optional*, defaults to "HuggingFace Transformers"):
```
Comment (sgugger) [inline src/transformers/integrations.py:1313]:
```suggestion
            CLEARML_PROJECT (`str`, *optional*, defaults to `"HuggingFace Transformers"`):
```

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -1299,6 +1305,114 @@ def on_train_end(self, args, state, control, **kwargs):
...
+                ClearML project name. If not set deafault project name will be assigned.
```
Comment (sgugger) [inline src/transformers/integrations.py:1314]:
```suggestion
                ClearML project name.
```

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -1299,6 +1305,114 @@ def on_train_end(self, args, state, control, **kwargs):
...
+            CLEARML_TASK (`str`, *optional* defaults to `"Trainer"`):
+                ClearML task name. If not set default task name will be assigned.
```
Comment (sgugger) [inline src/transformers/integrations.py:1316]:
```suggestion
                ClearML task name.
```

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -1299,6 +1305,114 @@ def on_train_end(self, args, state, control, **kwargs):
...
+    """
+
+    def __init__(self):
+        has_clearml = is_clearml_available()
+
+        if has_clearml:
```
Comment (sgugger) [inline src/transformers/integrations.py:1322]:
```suggestion
        if is_clearml_available():
```
I don't think you need to store this result ;-)

**Review by sgugger — COMMENTED** [review]
Thanks for working on adding this! It would be great to also add `clearml` to the documentation of the [`report_to`](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.report_to) training argument (so that users know it can be used). You should also add your new callback to [this doc page](https://huggingface.co/docs/transformers/main_classes/callback)!

Also don't forget to run `make style` on your branch so the formatting of your code is done :-)

**Comment by skinan** [comment]
@sgugger , We have made the changes according to your suggestions. Please have a look. Thank you for your co-operation.

**Review by sgugger — APPROVED** [review]
Thanks for iterating on this!

**Comment by pasky** [comment]
I'm finding this integration very aggressive.
Basically, it seems to require that if the clearml python package is installed, you must have clearml set up in order to use huggingface.

We have optional clearml setup in our environment (with a private clearml server), but not using it in most situations at the moment.

This now started crashing with
```
File /usr/local/lib/python3.8/dist-packages/clearml/backend_api/session/session.py:180, in Session.__init__(self, worker, api_key, secret_key, host, logger, verbose, initialize_logging, config, http_retries_config, **kwargs)
    177     raise ValueError("ClearML host was not set, check your configuration file or environment variable")
    179 if not self._offline_mode and (not self.secret_key and not self.access_key and not self.__auth_token):
--> 180     raise MissingConfigError()
    182 self._ssl_error_count_verbosity = self.config.get(
    183     "api.ssl_error_count_verbosity", self._ssl_error_count_verbosity)
    185 self.__host = host.strip("/")

MissingConfigError: It seems ClearML is not configured on this machine!
To get started with ClearML, setup your own 'clearml-server' or create a free account at https://app.clear.ml/
Setup instructions can be found here: https://clear.ml/docs
```

**Comment by sgugger** [comment]
You should have seen a log informing you that `report_to` was defaulting to all tools installed (issued [here](https://github.com/huggingface/transformers/blob/3830b3f74a57b3bcf6f14016c483fb0bb14b01ce/src/transformers/training_args.py#L1218)). We should probably change it to a warning so that it's more visible, but the default will change in v5 of Transformers (from all to None).

**Pull request merged**